src/accumulator/stump.rs

```diff
@@ -100,53 +117,67 @@ impl Stump {
         proof: &Proof,
-    ) -> Result<Vec<bitcoin_hashes::sha256::Hash>, String> {
+    ) -> Result<(Vec<(u64, sha256::Hash)>, Vec<sha256::Hash>), String> {
         if del_hashes.len() == 0 {
-            return Ok(self.roots.clone());
+            return Ok((vec![], self.roots.clone()));
         }
 
         let del_hashes = vec![sha256::Hash::default(); proof.targets()];
-        let (_, new_roots) = proof.calculate_hashes(&del_hashes, self)?;
-
-        Ok(new_roots)
+        proof.calculate_hashes(&del_hashes, self)
     }
     /// Adds new leafs into the root
     fn add(
         mut roots: Vec<bitcoin_hashes::sha256::Hash>,
         utxos: &Vec<bitcoin_hashes::sha256::Hash>,
-        mut leafs: u64,
-    ) -> Vec<bitcoin_hashes::sha256::Hash> {
-        for i in utxos.iter() {
-            Stump::add_single(&mut roots, *i, leafs);
-            leafs += 1;
-        }
+        mut leaves: u64,
+    ) -> (Vec<sha256::Hash>, Vec<(u64, sha256::Hash)>, Vec<u64>) {
+        let after_rows = util::tree_rows(leaves + (utxos.len() as u64));
+        let mut updated_subtree: Vec<(u64, sha256::Hash)> = vec![];
+        let all_deleted = util::roots_to_destroy(utxos.len() as u64, leaves, &roots);
+
+        for (i, add) in utxos.iter().enumerate() {
+            let mut pos = leaves;
+
+            // deleted is the empty roots that are being added over. These force
+            // the current root to move up.
+            let deleted = util::roots_to_destroy((utxos.len() - i) as u64, leaves, &roots);
+            for del in deleted {
+                if util::is_ancestor(util::parent(del, after_rows), pos, after_rows).unwrap() {
+                    pos = util::calc_next_pos(pos, del, after_rows).unwrap();
+                }
+            }
+            let mut h = 0;
+            // Iterates over roots, if we find a root that is not empty, we concatenate with
+            // the one we are adding and create new root, leaving this position empty. Stops
+            // when find an empty root.
+
+            // You can say if a root is empty, by looking a the binary representations of the
+            // number of leafs. If the h'th bit is one, then this position is occupied, empty
+            // otherwise.
+            let mut to_add = add.clone();
+            while (leaves >> h) & 1 == 1 {
+                let root = roots.pop();
+
+                if let Some(root) = root {
+                    if root != sha256::Hash::default() {
+                        updated_subtree.push((util::left_sibling(pos), root));
+                        updated_subtree.push((pos, to_add));
+                        pos = util::parent(pos, after_rows);
+
+                        to_add = types::parent_hash(&root, &to_add);
+                    }
+                }
+                h += 1;
+            }
+            updated_subtree.push((pos, to_add));
 
-        roots
-    }
+            updated_subtree.sort();
```

Comment:
In the go one I had a vector of subtrees because you could avoid the sort & dedupe here.

Maybe in rust those operations aren't as slow but just noting here for future reference.